### PR TITLE
feat: add ability to write Range values with JSONStreamWriter

### LINKED ISSUE #2481: Add support for Range types

### PR DESCRIPTION
**Pull request by PhongChuong**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/java-bigquerystorage/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Fixes #2481  ☕️

### CODE REVIEW

File: google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1/it/ITBigQueryStorageTest.java [hunk 1]
```diff
@@ -329,6 +399,125 @@ public void testRangeType() throws InterruptedException {
...
+    String streamName = session.getStreams(0).getName();
+
+    ReadRowsRequest readRowsRequest =
+        ReadRowsRequest.newBuilder().setReadStream(streamName).build();
+
+    long rowCount = 0;
+    ServerStream<ReadRowsResponse> stream = client.readRowsCallable().call(readRowsRequest);
+    for (ReadRowsResponse response : stream) {
+      Preconditions.checkState(response.hasArrowRecordBatch());
+      rowCount += response.getRowCount();
```
Comment (yirutang) [inline google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1/it/ITBigQueryStorageTest.java:516]: validate the value?

File: google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1/it/ITBigQueryStorageTest.java [hunk 2]
```diff
@@ -329,6 +399,125 @@ public void testRangeType() throws InterruptedException {
...
+      rangeDatetime2.put("start", "2014-08-19T05:41:35.220000");
+      row2.put("datetime", rangeDatetime1);
+      JSONObject rangeTimestamp2 = new JSONObject();
+      rangeTimestamp2.put("start", 1715360343);
+      row2.put("timestamp", rangeTimestamp1);
+      data.put(row2);
+
+      JSONObject row3 = new JSONObject();
+      JSONObject rangeDate3 = new JSONObject();
+      rangeDate3.put("start", 18262);
```
Comment (yirutang) [inline google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1/it/ITBigQueryStorageTest.java:461]: test date as a string?

File: google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1/JsonToProtoMessageTest.java [hunk 3]
```diff
@@ -1019,6 +1019,76 @@ public void testRequired() throws Exception {
...
+    JSONArray data = new JSONArray();
+    JSONObject row = new JSONObject();
+
+    JSONObject rangeDate = new JSONObject();
+    rangeDate.put("start", 18262);
+    rangeDate.put("end", 18627);
+    row.put("range_date", rangeDate);
+
+    JSONObject rangeDatetime = new JSONObject();
+    rangeDatetime.put("start", 1715360343);
```
Comment (yirutang) [inline google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1/JsonToProtoMessageTest.java:1076]: I think string inputs on these values will be more common.

**Review by yirutang — COMMENTED** [review]
Looks good! Some nits.

File: google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1/JsonToProtoMessageTest.java [hunk 4]
```diff
@@ -1019,6 +1019,76 @@ public void testRequired() throws Exception {
...
+                TestRangeTimestamp.newBuilder().setStart(1715360343).setEnd(1715446743))
+            .build();
+
+    JSONArray data = new JSONArray();
+    JSONObject row = new JSONObject();
+
+    JSONObject rangeDate = new JSONObject();
+    rangeDate.put("start", 18262);
+    rangeDate.put("end", 18627);
+    row.put("range_date", rangeDate);
```
Comment (yirutang) [inline google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1/JsonToProtoMessageTest.java:1073]: Do we require cases to match here? 
We should at least test field name with different case? 

Can "start" and "end" be case insensitive? This is more like an open question.

File: google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1/it/ITBigQueryStorageTest.java (hunk 1 above)
Comment (PhongChuong) [reply google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1/it/ITBigQueryStorageTest.java]: Added validation.

**Review by PhongChuong — COMMENTED** [review]
@yirutang 
Thanks for the review. I've updated the PR with the suggestions. PTAL.

File: google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1/JsonToProtoMessageTest.java [hunk 5]
```diff
@@ -1019,6 +1019,128 @@ public void testRequired() throws Exception {
...
+
+    JSONObject rangeTimestamp = new JSONObject();
+    rangeTimestamp.put("start", 1715360343);
+    rangeTimestamp.put("end", 1715446743);
+    row.put("range_timestamp", rangeTimestamp);
+
+    JSONObject rangeDateMixedCase = new JSONObject();
+    rangeDateMixedCase.put("START", "1970-01-02");
+    rangeDateMixedCase.put("eND", "1970-01-03");
+    row.put("range_date_mixed_case", rangeDateMixedCase);
```
Comment (yirutang) [inline google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1/JsonToProtoMessageTest.java:1125]: The field name itself should be case insensitive.

File: google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1/it/ITBigQueryStorageTest.java [hunk 6]
```diff
@@ -329,6 +607,130 @@ public void testRangeType() throws InterruptedException {
...
+
+        JSONObject datetimeColumn = new JSONObject();
+        Range datetime = RANGE_TEST_VALUES_DATETIME.get(name);
+        if ((!datetime.getStart().isNull()) && (datetime.getStart().getStringValue() != null)) {
+          datetimeColumn.put("start", datetime.getStart().getStringValue());
+        }
+        if ((!datetime.getEnd().isNull()) && (datetime.getEnd().getStringValue() != null)) {
+          datetimeColumn.put("end", datetime.getEnd().getStringValue());
+        }
+        row.put("datetime", datetimeColumn);
```
Comment (yirutang) [inline google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1/it/ITBigQueryStorageTest.java:655]: Try to use "dateTime" as field name instead and it should be able to pass.

File: google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1/JsonToProtoMessageTest.java (hunk 5 above)
Comment (PhongChuong) [reply google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1/JsonToProtoMessageTest.java:1125]: Updated.

File: google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1/it/ITBigQueryStorageTest.java (hunk 6 above)
Comment (PhongChuong) [reply google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1/it/ITBigQueryStorageTest.java:655]: Updated.

File: google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1/JsonToProtoMessageTest.java (hunk 5 above)
Comment (yirutang) [reply google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1/JsonToProtoMessageTest.java:1125]: Where is the update? I mean to have name like this "Range_Date_Mixed_Case"

**Review by yirutang — APPROVED** [review]

**Pull request merged**